src/Interpreters/Context.h

```diff
@@ -716,8 +716,15 @@ class Context: public ContextData, public std::enable_shared_from_this<Context>
     };
 
     String resolveDatabase(const String & database_name) const;
-    StorageID resolveStorageID(StorageID storage_id, StorageNamespace where = StorageNamespace::ResolveAll) const;
-    StorageID tryResolveStorageID(StorageID storage_id, StorageNamespace where = StorageNamespace::ResolveAll) const;
+
+    /// Find table.
+    /// If the table is in a regular database (not external, not temporary):
+    ///  * If get_uuid is true, do the lookup in the database, and if the table is found return StorageID with UUID.
+    ///    Use this if it's preferable for the returned StorageID to identify a specific table.
+    ///  * If `get_uuid` is false, a StorageID without UUID will be returned.
+    ///    Use this if you're going to do a lookup in the database right away anyway.
+    StorageID resolveStorageID(StorageID storage_id, StorageNamespace where = StorageNamespace::ResolveAll, bool get_uuid = true) const;
+    StorageID tryResolveStorageID(StorageID storage_id, StorageNamespace where = StorageNamespace::ResolveAll, bool get_uuid = true) const;
```

Comment:
Why do we need `get_uuid = false`? The idea was: resolve UUIDs at the beginning of query execution, access tables by UUID, so we don't care about names that can change

Reply:
The scenario this PR is preventing is that `resolveStorageID` gets UUID, then the table gets exchanged+dropped, then we fail to `lockForShare` it because it's dropped. If the {UUID lookup, `lockForShare`} pair of operations were more like a transaction then this failure mode doesn't exist. We would either find+lock a non-dropped table (success) or we would see a moment in time when no non-dropped table with name existed (UNKNOWN_TABLE). No weird case where a readable table with this name exists at every moment in time but we failed read from any such table.

Reply:
Ah, sorry, I misunderstood the problem. Maybe we can change the behavior of `lockForShare` for "dropped" tables, so it will not throw, but I'm not sure. But taking the locks while resolving UUIDs makes sense too